File: dotnet/tests/ApplePurchaseReceiptVerifier.Tests/ReceiptBase64FastPathTests.cs
using System;
using System.Globalization;
using System.Text;
using ApplePurchaseReceiptVerifier.Receipt;
using Xunit;

namespace ApplePurchaseReceiptVerifier.Tests;

/// <summary>
/// <c>ReceiptVerifier.DecodeBase64</c> tries <c>Convert.TryFromBase64String</c>
/// before its own tolerant parser. That shortcut is only safe if it never
/// changes an answer: every string the fast path accepts must be one the
/// tolerant parser accepts too, with the same bytes, and every string it
/// declines must reach the tolerant parser unchanged. If the fast path ever
/// accepted something the documented <c>receipt-data</c> contract rejects (an
/// empty string, a character after the padding, a wrong padding count), a
/// malformed receipt would start verifying in .NET alone and the ports would
/// disagree.
/// </summary>
/// <remarks>
/// <para><c>TryFromBase64String</c> is not strict: it skips whitespace
/// anywhere in the string. That is why this is a differential test rather
/// than an argument, and why it runs on every runtime the suite targets
/// (net8.0, net9.0, net10.0): the decoder's rules belong to the runtime, not
/// to this library.</para>
/// <para>The generator is seeded, so a failure reproduces. It mixes the
/// shapes the contract names: clean standard base64, missing and extra
/// padding, both alphabets, whitespace anywhere, illegal characters (Unicode
/// whitespace among them), characters after the padding, and lengths
/// congruent to 1 mod 4.</para>
/// </remarks>
public class ReceiptBase64FastPathTests
{
    private const int Seed = 0x5EEDB64;
    private const int Cases = 20_000;
    private const string Illegal =
        "!#$%&*.,:;?@[]{}|~\"'\\\u0000\u007f\u00e9\u00ff\u0100\u00a0\u0085\u000b\u000c\u2028\u3000\uff21";
    private const string Whitespace = "\r\n \t";

    [Fact]
    public void DecodeAgreesWithTheTolerantPathOnGeneratedInputs()
    {
        Random random = new Random(Seed);
        int fastAccepted = 0;
        int fastAcceptedWithWhitespace = 0;
        int tolerantOnlyAccepted = 0;
        int rejected = 0;
        for (int i = 0; i < Cases; i++)
        {
            string input = Generate(random);
            AssertAgrees(input);
            if (ReceiptVerifier.DecodeBase64Fast(input) is not null)
            {
                fastAccepted++;
                if (input.IndexOfAny(Whitespace.ToCharArray()) >= 0)
                {
                    fastAcceptedWithWhitespace++;
                }
            }
            else if (Accepts(ReceiptVerifier.DecodeBase64Tolerant, input))
            {
                tolerantOnlyAccepted++;
            }
            else
            {
                rejected++;
            }
        }

        // The comparison proves nothing unless every branch ran many times:
        // the fast path taken (with and without the whitespace it skips), the
        // fallback accepting, and both rejecting.
        Assert.True(fastAccepted > 1_000, "fast path accepted only " + fastAccepted);
        Assert.True(
            fastAcceptedWithWhitespace > 100,
            "fast path accepted whitespace only " + fastAcceptedWithWhitespace);
        Assert.True(tolerantOnlyAccepted > 1_000, "fallback accepted only " + tolerantOnlyAccepted);
        Assert.True(rejected > 1_000, "rejected only " + rejected);
    }

    [Theory]
    [InlineData("")]
    [InlineData(" ")]
    [InlineData("\r\n\t ")]
    [InlineData("=")]
    [InlineData("==")]
    [InlineData("===")]
    [InlineData("====")]
    [InlineData("A")]
    [InlineData("A=")]
    [InlineData("A==")]
    [InlineData("A===")]
    [InlineData("AA")]
    [InlineData("AA=")]
    [InlineData("AA==")]
    [InlineData("AA===")]
    [InlineData("AA= =")]
    [InlineData("AAA")]
    [InlineData("AAA=")]
    [InlineData("AAA==")]
    [InlineData("AAAA")]
    [InlineData("AAAA=")]
    [InlineData("AAAA==")]
    [InlineData("AAAA====")]
    [InlineData("AA==A")]
    [InlineData("AA==\n")]
    [InlineData("AA==AA==")]
    [InlineData("AAAAA")]
    [InlineData("-_")]
    [InlineData("+/")]
    [InlineData("+_")]
    [InlineData("-/")]
    [InlineData("AB-_")]
    [InlineData("AB+/")]
    [InlineData("AB\u00e9=")]
    [InlineData("AB\u0100")]
    [InlineData("AB\u00a0CD")]
    [InlineData("AB\u000bCD")]
    [InlineData("QUJD")]
    [InlineData("QUJ")]
    [InlineData("QUI")]
    [InlineData("QQ")]
    [InlineData("QR")]
    [InlineData("QUK=")]
    [InlineData("QUJD\r\n")]
    [InlineData("Q U J D")]
    public void DecodeAgreesWithTheTolerantPathOnHandPickedEdges(string input)
    {
        AssertAgrees(input);
    }

    /// <summary>
    /// The comparison has teeth: the obvious fast path without the zero-byte
    /// refusal accepts the empty string, and a permissive one that also takes
    /// base64url mixes the alphabets. Both would change answers, and both are
    /// caught by the same seeded run that the real fast path passes.
    /// </summary>
    [Fact]
    public void AFastPathThatChangesAnAnswerFailsTheComparison()
    {
        Func<string, byte[]?> unguarded = s =>
        {
            byte[] buffer = new byte[s.Length];
            return Convert.TryFromBase64String(s, buffer, out int written) ? buffer.AsSpan(0, written).ToArray() : null;
        };
        Func<string, byte[]?> permissive = s =>
        {
            byte[] buffer = new byte[s.Length];
            string translated = s.Replace('-', '+').Replace('_', '/');
            return Convert.TryFromBase64String(translated, buffer, out int written) && written > 0
                ? buffer.AsSpan(0, written).ToArray()
                : null;
        };

        Assert.True(FirstDisagreement(unguarded) is not null, "the unguarded decoder was not caught");
        Assert.True(FirstDisagreement(permissive) is not null, "the permissive decoder was not caught");
        Assert.Null(FirstDisagreement(ReceiptVerifier.DecodeBase64Fast));
    }

    private static string? FirstDisagreement(Func<string, byte[]?> fastPath)
    {
        Random random = new Random(Seed);
        for (int i = 0; i < Cases; i++)
        {
            string input = Generate(random);
            string? problem = Disagreement(s => fastPath(s) ?? ReceiptVerifier.DecodeBase64Tolerant(s), input);
            if (problem is not null)
            {
                return problem;
            }
        }

        return null;
    }

    private static void AssertAgrees(string input)
    {
        string? problem = Disagreement(ReceiptVerifier.DecodeBase64, input);
        Assert.True(problem is null, problem);
    }

    /// <summary>
    /// Null when <paramref name="decode"/> and the tolerant path give the same
    /// answer for <paramref name="input"/> (equal bytes, or the same reason
    /// and message); otherwise what differs.
    /// </summary>
    private static string? Disagreement(Func<string, byte[]> decode, string input)
    {
        byte[]? expected = null;
        VerificationException? expectedError = null;
        try
        {
            expected = ReceiptVerifier.DecodeBase64Tolerant(input);
        }
        catch (VerificationException e)
        {
            expectedError = e;
        }

        byte[]? actual = null;
        VerificationException? actualError = null;
        try
        {
            actual = decode(input);
        }
        catch (VerificationException e)
        {
            actualError = e;
        }

        string shown = Describe(input);
        if (expectedError is not null)
        {
            if (actualError is null)
            {
                return "accepted what the tolerant path rejects: " + shown;
            }

            return expectedError.Reason == actualError.Reason
                && string.Equals(expectedError.Message, actualError.Message, StringComparison.Ordinal)
                ? null
                : "rejected differently: " + shown + ": " + actualError.Message;
        }

        if (actualError is not null)
        {
            return "rejected what the tolerant path accepts: " + shown + ": " + actualError.Message;
        }

        return expected.AsSpan().SequenceEqual(actual) ? null : "decoded different bytes: " + shown;
    }

    private static bool Accepts(Func<string, byte[]> decode, string input)
    {
        try
        {
            decode(input);
            return true;
        }
        catch (VerificationException)
        {
            return false;
        }
    }

    private static string Generate(Random random)
    {
        byte[] bytes = new byte[random.Next(40)];
        random.NextBytes(bytes);
        StringBuilder s = new StringBuilder(Convert.ToBase64String(bytes));

        // About half the inputs stay canonical standard base64 so the fast
        // path is taken often; the rest get one or more defects.
        if (random.Next(2) == 0)
        {
            return s.ToString();
        }

        int mutations = 1 + random.Next(3);
        for (int m = 0; m < mutations; m++)
        {
            switch (random.Next(9))
            {
                case 0: // drop the padding
                    TrimPadding(s);
                    break;
                case 1: // extra padding
                    s.Append(random.Next(2) == 0 ? "=" : "==");
                    break;
                case 2: // base64url alphabet, whole string
                    s.Replace('+', '-').Replace('/', '_');
                    break;
                case 3: // one character of either alphabet's extras
                    Insert(s, random, "+/-_"[random.Next(4)]);
                    break;
                case 4: // whitespace somewhere, including before and after the padding
                    Insert(s, random, Whitespace[random.Next(Whitespace.Length)]);
                    break;
                case 5: // an illegal character
                    Insert(s, random, Illegal[random.Next(Illegal.Length)]);
                    break;
                case 6: // something after the padding
                    s.Append('=').Append((char)('A' + random.Next(26)));
                    break;
                case 7: // length congruent to 1 mod 4 (after stripping padding)
                    TrimPadding(s);
                    while (s.Length % 4 != 1)
                    {
                        s.Append('A');
                    }

                    break;
                default: // whitespace-only or empty, or trailing whitespace
                    if (random.Next(8) == 0)
                    {
                        s.Clear();
                        for (int k = random.Next(3); k > 0; k--)
                        {
                            s.Append(Whitespace[random.Next(Whitespace.Length)]);
                        }
                    }
                    else
                    {
                        s.Append(Whitespace[random.Next(Whitespace.Length)]);
                    }

                    break;
            }
        }

        return s.ToString();
    }

    private static void TrimPadding(StringBuilder s)
    {
        while (s.Length > 0 && s[s.Length - 1] == '=')
        {
            s.Length--;
        }
    }

    private static void Insert(StringBuilder s, Random random, char c)
    {
        s.Insert(random.Next(s.Length + 1), c);
    }

    private static string Describe(string input)
    {
        StringBuilder output = new StringBuilder("\"");
        foreach (char c in input)
        {
            if (c >= 0x20 && c < 0x7f)
            {
                output.Append(c);
            }
            else
            {
                output.Append("\\u").Append(((int)c).ToString("x4", CultureInfo.InvariantCulture));
            }
        }

        return output.Append('"').ToString();
    }
}

File: swift/Tests/ApplePurchaseReceiptVerifierTests/ReceiptBase64FastPathTests.swift
import Foundation
import XCTest

@testable import ApplePurchaseReceiptVerifier

/// `decodeReceiptBase64` tries Foundation's decoder before its own tolerant
/// parser. That shortcut is only safe if it never changes an answer: every
/// string it accepts must be one the tolerant parser accepts too, with the
/// same bytes, and every string it refuses must reach the tolerant parser
/// unchanged. If the fast path ever accepted something the documented
/// `receipt-data` contract rejects (a character after the padding, a wrong
/// padding count), a malformed receipt would start verifying in Swift alone,
/// and the ports would disagree.
///
/// This matters more here than in the other ports because Foundation's
/// decoder is not the same code on every platform: on Linux it accepts
/// over-padded strings such as "AAAA=" that the contract rejects, which is
/// why the fast path checks the shape itself first. These tests run on
/// whichever Foundation the suite is built against, so CI proves it on Linux
/// and on macOS separately.
///
/// The generator is seeded, so a failure reproduces. It mixes the shapes the
/// contract names: clean standard base64, missing and extra padding, both
/// alphabets, whitespace anywhere, illegal characters (including ones outside
/// Latin-1), characters after the padding, and lengths congruent to 1 mod 4.
final class ReceiptBase64FastPathTests: XCTestCase {
    private static let seed: UInt64 = 0x5EED_B64
    private static let cases = 20_000
    private static let illegal = Array("!#$%&*.,:;?@[]{}|~\"'\\\u{0000}\u{007F}éÿĀ\u{00A0}\u{2028}\u{000B}")
    private static let whitespace: [Character] = ["\r", "\n", " ", "\t"]

    func testDecodeAgreesWithTheTolerantPathOnGeneratedInputs() {
        var random = SplitMix64(seed: Self.seed)
        var fastAccepted = 0
        var tolerantOnlyAccepted = 0
        var rejected = 0
        for _ in 0..<Self.cases {
            let input = generate(&random)
            if compare(input) {
                if fastPathAccepts(input) {
                    fastAccepted += 1
                } else {
                    tolerantOnlyAccepted += 1
                }
            } else {
                rejected += 1
            }
        }
        // The comparison proves nothing unless every branch was exercised
        // many times: the fast path taken, the fallback accepting, and both
        // rejecting.
        XCTAssertGreaterThan(fastAccepted, 1_000, "fast path accepted only \(fastAccepted)")
        XCTAssertGreaterThan(tolerantOnlyAccepted, 1_000, "fallback accepted only \(tolerantOnlyAccepted)")
        XCTAssertGreaterThan(rejected, 1_000, "rejected only \(rejected)")
    }

    func testDecodeAgreesWithTheTolerantPathOnHandPickedEdges() {
        let edges = [
            "", " ", "\r\n\t ", "=", "==", "===", "====",
            "A", "A=", "A==", "A===", "AA", "AA=", "AA==", "AA===",
            "AAA", "AAA=", "AAA==", "AAAA", "AAAA=", "AAAA==", "AAAA====",
            "AA==A", "AA==\n", "AA==AA==", "AAAAA",
            "-_", "+/", "+_", "-/", "AB-_", "AB+/", "ABé=", "ABĀ",
            "QUJD", "QUJ", "QUI", "QQ", "QR", "QR==", "QUK=", "QUJD\r\n",
        ]
        for edge in edges {
            compare(edge)
        }
    }

    /// The shape check is what keeps Foundation's leniency out: each of these
    /// is a string Foundation on Linux decodes and the contract rejects, so
    /// the guard must refuse it before the decoder runs.
    func testShapeCheckRefusesWhatFoundationWouldOverAccept() {
        for input in ["AA===", "AAAA=", "AAAA==", "AAAA====", "A===", "===="] {
            XCTAssertFalse(isCanonicalStandardBase64(input), input.debugDescription)
            XCTAssertNil(decodeReceiptBase64(input), input.debugDescription)
        }
    }

    /// Asserts that `decodeReceiptBase64` and `decodeReceiptBase64Tolerant`
    /// give the same answer for `input`: equal bytes, or both nil. Returns
    /// whether the input was accepted. The Swift decoder reports a rejection
    /// as nil, with no reason or message of its own; the callers turn every
    /// nil into the same `.invalidReceiptFormat` error.
    @discardableResult
    private func compare(_ input: String) -> Bool {
        let expected = decodeReceiptBase64Tolerant(input)
        let actual = decodeReceiptBase64(input)
        let shown = input.debugDescription
        guard let expected else {
            XCTAssertNil(actual, "decode accepted what the tolerant path rejects: \(shown)")
            return false
        }
        XCTAssertEqual(actual, expected, shown)
        return true
    }

    private func fastPathAccepts(_ input: String) -> Bool {
        isCanonicalStandardBase64(input) && Data(base64Encoded: input) != nil
    }

    private func generate(_ random: inout SplitMix64) -> String {
        let bytes = (0..<Int.random(in: 0..<40, using: &random)).map { _ in
            UInt8.random(in: 0...255, using: &random)
        }
        var s = Array(Data(bytes).base64EncodedString())
        // About half the inputs stay canonical standard base64 so the fast
        // path is taken often; the rest get one or more defects.
        if Bool.random(using: &random) {
            return String(s)
        }
        for _ in 0..<Int.random(in: 1...3, using: &random) {
            switch Int.random(in: 0..<9, using: &random) {
            case 0:  // drop the padding
                while s.last == "=" { s.removeLast() }
            case 1:  // extra padding
                s.append(contentsOf: Bool.random(using: &random) ? "=" : "==")
            case 2:  // base64url alphabet, whole string
                s = s.map { $0 == "+" ? "-" : $0 == "/" ? "_" : $0 }
            case 3:  // one character of the other alphabet
                insert(&s, Array("+/-_").randomElement(using: &random)!, &random)
            case 4:  // whitespace somewhere, including before and after the padding
                insert(&s, Self.whitespace.randomElement(using: &random)!, &random)
            case 5:  // an illegal character
                insert(&s, Self.illegal.randomElement(using: &random)!, &random)
            case 6:  // something after the padding
                s.append("=")
                s.append(Character(Unicode.Scalar(UInt8(ascii: "A") + UInt8.random(in: 0..<26, using: &random))))
            case 7:  // length congruent to 1 mod 4 (after stripping padding)
                while s.last == "=" { s.removeLast() }
                while s.count % 4 != 1 { s.append("A") }
            default:  // whitespace-only or empty
                if Int.random(in: 0..<8, using: &random) == 0 {
                    s = (0..<Int.random(in: 0..<3, using: &random)).map { _ in
                        Self.whitespace.randomElement(using: &random)!
                    }
                } else {
                    s.append(Self.whitespace.randomElement(using: &random)!)
                }
            }
        }
        // String(s) joins "\r" and "\n" into one Character, but the decoders
        // walk UTF-8 bytes, so the joined form is what a client would send.
        return String(s)
    }

    private func insert(_ s: inout [Character], _ c: Character, _ random: inout SplitMix64) {
        s.insert(c, at: Int.random(in: 0...s.count, using: &random))
    }
}

/// A seeded generator, so a failing input can be reproduced. The standard
/// library's `SystemRandomNumberGenerator` cannot be seeded.
struct SplitMix64: RandomNumberGenerator {
    private var state: UInt64

    init(seed: UInt64) {
        state = seed
    }

    mutating func next() -> UInt64 {
        state &+= 0x9E37_79B9_7F4A_7C15
        var z = state
        z = (z ^ (z >> 30)) &* 0xBF58_476D_1CE4_E5B9
        z = (z ^ (z >> 27)) &* 0x94D0_49BB_1331_11EB
        return z ^ (z >> 31)
    }
}
